### PR TITLE
TC to validate the restores for multiple backups when intermittent backups are deleted

### PR DESCRIPTION
**Pull request by sabrarhussaini**
Added a new TC **IssueMultipleBackupsAndRestoreInterleavedCopies** to verify the restores for multiple backups when intermittent backups are deleted.

Below are the TC details:
Task: https://portworx.atlassian.net/browse/PB-5385
Testrail: https://portworx.testrail.net/index.php?/cases/view/58049

Which issue(s) this PR fixes (optional)
Closes #PB-5385

Special notes for your reviewer:
Vanilla-S3:
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/4430/

Vanilla NFS:
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/4431/

IKS-S3:
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/Cloud_BYOC/91/
https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/Cloud_BYOC/92/

Note: Having go routines for backup creation wasn't helpful as the backup completed weren't sequential.

### CODE REVIEW

File: tests/backup/backup_kdmp_test.go [hunk 1]
```diff
@@ -1400,3 +1401,158 @@ var _ = Describe("{ExcludeInvalidDirectoryFileBackup}", func() {
...
+				backupNameList = append(backupNameList, currentBackupName)
+			}
+			log.Infof("List of backups - %v", backupNameList)
+		})
+
+		Step("Deleting few intermittent backups", func() {
+			log.InfoD("Deleting %v random intermittent backups", backupDeleteCount)
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			rand.Seed(time.Now().UnixNano())
```
Comment (snigdha-px) [inline tests/backup/backup_kdmp_test.go:1501]: what is the use of above line?

**Review by snigdha-px — APPROVED** [review]
LGTM except one minor query for my understanding

File: tests/backup/backup_kdmp_test.go [hunk 2]
```diff
@@ -1400,3 +1401,158 @@ var _ = Describe("{ExcludeInvalidDirectoryFileBackup}", func() {
...
+		log.InfoD("Deleting the deployed apps after the testcase")
+		opts := make(map[string]bool)
+		opts[SkipClusterScopedObjects] = true
+		DestroyApps(scheduledAppContexts, opts)
+		log.InfoD("Deleting the px-backup objects")
+		CleanupCloudSettingsAndClusters(backupLocationMap, cloudAccountName, cloudCredUID, ctx)
+		log.InfoD("Switching context to destination cluster for clean up")
+		err = SetDestinationKubeConfig()
+		log.FailOnError(err, "Unable to switch context to destination cluster [%s]", destinationClusterName)
+		DestroyApps(scheduledAppContexts, opts)
```
Comment (sn-px) [inline tests/backup/backup_kdmp_test.go:1553]: Is is this required as you are not deploying any application on destination cluster

File: tests/backup/backup_kdmp_test.go [hunk 3]
```diff
@@ -1400,3 +1401,158 @@ var _ = Describe("{ExcludeInvalidDirectoryFileBackup}", func() {
...
+			ValidateApplications(scheduledAppContexts)
+		})
+
+		Step("Validate creation of cloud credentials and backup location", func() {
+			log.InfoD("Validate creation of cloud credentials and backup location")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			for _, provider := range providers {
+				cloudCredUID = uuid.New()
+				cloudAccountName = fmt.Sprintf("%s-%s-%v", "cred", provider, RandomString(4))
```
Comment (sn-px) [inline tests/backup/backup_kdmp_test.go:1456]: create constant CredName here instead of "cred"

**Review by sn-px — CHANGES_REQUESTED** [review]

File: tests/backup/backup_kdmp_test.go (hunk 2 above)
Comment (sabrarhussaini) [reply tests/backup/backup_kdmp_test.go:1553]: Kindly correct me if I'm wrong, but we do restore the applications on the destination cluster which needs to be cleaned up.

File: tests/backup/backup_kdmp_test.go (hunk 3 above)
Comment (sabrarhussaini) [reply tests/backup/backup_kdmp_test.go:1456]: @sn-px I'm sorry, I couldn't comprehend the comment. Can you please let me know what exact value should be considered? now the format for the cloudAccountName will be "cred-aws-xxxx", did you mean it should be instead "aws-CredName-xxxx  "

Comment (sabrarhussaini) [reply tests/backup/backup_kdmp_test.go]: @sn-px Addressed the comment.

**Review by sn-px — CHANGES_REQUESTED** [review]

File: tests/backup/backup_kdmp_test.go [hunk 4]
```diff
@@ -1400,3 +1401,158 @@ var _ = Describe("{ExcludeInvalidDirectoryFileBackup}", func() {
...
+		log.InfoD("Deleting the deployed apps after the testcase")
+		opts := make(map[string]bool)
+		opts[SkipClusterScopedObjects] = true
+		DestroyApps(scheduledAppContexts, opts)
+		log.InfoD("Deleting the px-backup objects")
+		CleanupCloudSettingsAndClusters(backupLocationMap, cloudAccountName, cloudCredUID, ctx)
+		log.InfoD("Switching context to destination cluster for clean up")
+		err = SetDestinationKubeConfig()
+		log.FailOnError(err, "Unable to switch context to destination cluster [%s]", destinationClusterName)
+		DestroyApps(scheduledAppContexts, opts)
```
Comment (mkoppal-px) [inline tests/backup/backup_kdmp_test.go:1553]: We dont have to do this. In `EndPxBackupTorpedoTest` we have a function called `DeleteAllNamespacesCreatedByTestCase` which will cleanup all the apps.

File: tests/backup/backup_kdmp_test.go [hunk 5]
```diff
@@ -1400,3 +1401,158 @@ var _ = Describe("{ExcludeInvalidDirectoryFileBackup}", func() {
 
 	})
 })
+
+// This TC takes 30 generic backups, deletes intermittent backups and validates the restores
+var _ = Describe("{IssueGenericBackupsAndRestoreInterleavedCopies}", func() {
```
Comment (kshithijiyer-px) [inline tests/backup/backup_kdmp_test.go:1406]: Go ahead and update the image.

Comment (sabrarhussaini) [reply tests/backup/backup_kdmp_test.go:1406]: @kshithijiyer-px, sorry- image? you mean the TC?

Comment (kshithijiyer-px) [reply tests/backup/backup_kdmp_test.go:1406]: Yes the testcase.

**Review by kshithijiyer-px — APPROVED** [review]
LGTM

File: tests/backup/backup_restore_basic_test.go [hunk 6]
```diff
@@ -4112,3 +4113,173 @@ var _ = Describe("{KubeAndPxNamespacesSkipOnAllNSBackup}", func() {
...
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching [%s] cluster uid", SourceClusterName))
+		})
+
+		Step("Taking multiple backups of application on source cluster", func() {
+			log.InfoD("Taking %v multiple backups of application on source cluster", backupCount)
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			appContextsToBackup := FilterAppContextsByNamespace(scheduledAppContexts, bkpNamespaces)
+			log.InfoD("Taking Backup of application")
+			for i := 0; i < backupCount; i++ {
```
Comment (mkoppal-px) [inline tests/backup/backup_restore_basic_test.go:4222]: We can run this in a go routine

Comment (sabrarhussaini) [reply tests/backup/backup_restore_basic_test.go:4222]: @mkoppal-px I tried it, what happens with routine is that the backups are not getting completed sequentially, though they are triggered in sequence but they remain in pending and the completion is actually random. Please advise.

Comment (mkoppal-px) [reply tests/backup/backup_restore_basic_test.go:4223]: Ok, lets go ahead with the sequential implementation for now

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Review by sn-px — APPROVED** [review]
LGTM

**Pull request merged**